experimental/zerotrie/src/zerotrie.rs

```diff
@@ -0,0 +1,625 @@
+/// ```
+#[repr(transparent)]
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub struct ZeroTrieSimpleAscii<Store: ?Sized> {
+    pub(crate) store: Store,
+}
+
+/// A data structure that compactly maps from byte strings to integers.
+///
+/// For more information, see [`ZeroTrie`].
+///
+/// # Examples
+///
+/// ```
+/// use zerotrie::ZeroTriePerfectHash;
+/// use litemap::LiteMap;
+///
+/// let mut map = LiteMap::<&[u8], usize>::new_vec();
+/// map.insert("foo".as_bytes(), 1);
+/// map.insert("bår".as_bytes(), 2);
+/// map.insert("båzzøø".as_bytes(), 3);
+///
+/// let trie = ZeroTriePerfectHash::try_from(&map)?;
+///
+/// assert_eq!(trie.get("foo".as_bytes()), Some(1));
+/// assert_eq!(trie.get("bår".as_bytes()), Some(2));
+/// assert_eq!(trie.get("båzzøø".as_bytes()), Some(3));
+/// assert_eq!(trie.get("bazzoo".as_bytes()), None);
+///
+/// # Ok::<_, zerotrie::ZeroTrieError>(())
+/// ```
+#[repr(transparent)]
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub struct ZeroTriePerfectHash<Store: ?Sized> {
+    pub(crate) store: Store,
+}
+
+/// A data structure that maps from a large number of byte strings to integers.
+///
+/// For more information, see [`ZeroTrie`].
+#[repr(transparent)]
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub struct ZeroTrieExtendedCapacity<Store: ?Sized> {
+    pub(crate) store: Store,
+}
+
+macro_rules! impl_zerotrie_subtype {
+    ($name:ident, $variant:ident, $getter_fn:path, $iter_ty:ty, $iter_fn:path, $cnv_fn:path) => {
+        impl<Store> $name<Store> {
+            /// Wrap this specific ZeroTrie variant into a ZeroTrie.
+            #[inline]
+            pub const fn into_zerotrie(self) -> ZeroTrie<Store> {
+                ZeroTrie(ZeroTrieFlavor::$variant(self))
+            }
+            /// Create a trie directly from a store.
+            ///
+            /// If the store does not contain valid bytes, unexpected behavior may occur.
+            #[inline]
+            pub const fn from_store(store: Store) -> Self {
+                Self { store }
+            }
+            /// Takes the byte store from this trie.
+            #[inline]
+            pub fn take_store(self) -> Store {
+                self.store
+            }
+            /// Maps the store into another type.
+            #[inline]
+            pub fn map_store<X>(self, f: impl FnOnce(Store) -> X) -> $name<X> {
+                $name::<X>::from_store(f(self.store))
+            }
+            #[inline]
+            pub(crate) fn map_store_into_zerotrie<X>(self, f: impl FnOnce(Store) -> X) -> ZeroTrie<X> {
+                $name::<X>::from_store(f(self.store)).into_zerotrie()
+            }
+        }
+        impl<Store> $name<Store>
+        where
+        Store: AsRef<[u8]> + ?Sized,
+        {
+            /// Queries the trie for a string.
+            #[inline]
+            pub fn get<K>(&self, key: K) -> Option<usize> where K: AsRef<[u8]> {
+                // TODO: Should this be AsRef or Borrow?
```

Comment:
The problem with `Borrow` is that I can't transparently take both `[u8]` and `str` because there is no Borrow impl between them, but there is an AsRef impl between them. So if I use Borrow I need to make `get_str`.